server/appServices.js

```diff
@@ -138,6 +144,22 @@ async function watchKeyCloakSecrets(namespace, kubeclient) {
   secretStream.on('end', () => {
     watchKeyCloakSecrets(namespace, kubeclient);
   });
+
+  secretStream.on('close', () => {
+    retryWatch(namespace, kubeclient);
```

Comment:
Why do we do the retry only when the keycloak stream is closed? 
Can we take for granted that it will never happen that, for example, the datasync stream is closed while the keycloak one is not?

Reply:
If the Kubernetes API is restarted then we will 100% not be able to connect to *any* resource. If we do this on every resource then we would be trying to reconnect to the server multiple times in parallel which is not needed.

Reply:
Yes, but are we sure that API restart is the only event that could cause the stream to be closed?

Reply:
maybe not, but I don't think it really matters, because when reconnect, it will reconnect for all resources. The only danger here is that is it possible that for whatever reason, all the other connections are closed but the secret one is not.  I am not sure if there is such a case, but if this is a concern, a safer way could be that we also listen to the `close` event on all the resources, but in the function `retryWatch` function itself,  ignore some of them if they are emitted in a very short of time period (something like a `debounce` function).

Reply:
@wei-lee 👍  That was exactly my concern. We could have any stream closed, not just keycloak (not only all but keycloak, but any but keycloak)